active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,998 @@
+blocked from being used for stack deployment extension configurations.**
+
+The stacks service already sends a stack resource ID HTTP request header to the deployments service. This will be used
+to enable additional validation in the Deployment engine to enforce these expression rules. It will also enable the
+return of extension configurations in the Deployment GET response for API versions that support it. This header is only
+needed for the initial deployment PUT, as its value will be saved on the deployment metadata and this will be used to
+drive the API behavior. For deployments that are not initiated by the stack service, the configuration portion of the
+extension details will not be returned, as no additional validation on language expressions is done, which means we
+cannot guarantee compliance to the ARM RP RPC rule that a resource GET cannot return secret data.
+
+#### Stacks mode for Bicep files
+
+Because the stack service will enable additional validation and features in the deployment service, Bicep template
+authors need to see diagnostics that flag disallowed expressions ahead of time. With the current Bicep file
+architecture, this cannot be determined automatically. Bicep files can be used with or without stacks and modules bind
+together different Bicep files to create a deployment hierarchy. To enable the additional diagnostics, they will need to
+be explicitly enabled by the template author. Here is the design for how to enable this in a Bicep file via the Bicep
+linter:
+
+```json5
+{
+  "rules": {
+    "stacks-extensibility-compatibility": { // break this down into smaller rules as needed
+      "level": "warning",
+      // The default level across all files. Up to user to configure as an error or to turn off.
+      // Alternative to file overrides is to stick with in-file/line overrides.
+      "fileOverrides": {
+        "*.stack.bicep": "error",
+        "<glob>": "error"
+      }
+    }
+  }
+}
+```
+
+There will be no ARM template changes to drive this. The extra validation on the backend will take place if the
+deployment is sent from the Stacks service. The Bicep compilation step for Stacks CLI commands should run with this mode
+set to "error" regardless of user setting to prevent templates that will certainly fail from being sent to the API.
+
+#### Passing extension configurations to nested deployments
+
+Users will need to reuse extension configurations across multiple nested deployments. Because these are passed as
+deployment properties rather than as parameters, there needs to be a way to reference this configuration in the
+consuming deployment for the nested deployment.
+
+Here are some examples of passing a configuration down another level:
+
+```bicep
+extension kubernetes // this was provided a configuration through the parent deployment/parameters file
+
+// ✅ whole inheritance
+module foo 'foo.bicep' = {
+  extensionConfigs: {
```

Comment:
It would be good to discuss naming for `extensionConfigs` (both here, and the proposed function) - it makes sense, but feels quite verbose

Reply:
There was some back and forth on this early on. We ended up sticking with `extensionConfigs`. If I recall correctly, the reason was to keep it close to the request body format to avoid a layer of indirection, as resource definitions need to align with their PUT. @shenglol any thoughts? I know you were keen on keeping it `extensionConfigs` versus `extensions`.

Reply:
One implication with the renaming the function `extensionConfigs` to `extensions` is that it will make more sense for the function to return not just the `config` object but also the extension `name` and `version`. This might be helpful if the user wants to output the extension name / version:

```Bicep
output k8sVersion string = k8s.version
```

```JSON
{
  "outputs": {
    "k8sVersion": "[extensions('k8s').version]"
  }
}
```